### ISSUE TITLE
oc cluster up - enable

### ISSUE THREAD

**Issue by cmoulliard** [snowdrop/openshift-infra#106]
When we do a `up/down` of the ansible cluster playbook, then we get this error

```
TASK [cluster : Enable options] ***************************************************************************************************************************************************************************************************************
fatal: [192.168.99.50]: FAILED! => {"changed": true, "cmd": ["oc", "cluster", "up", "--enable=[automation-service-broker,", "centos-imagestreams,", "persistent-volumes,", "registry,", "rhel-imagestreams,", "router,", "service-catalog,", "template-service-broker,", "web-console]",
 "--write-config=true", "--public-hostname=192.168.99.50", "--routing-suffix=192.168.99.50.nip.io", "--server-loglevel=1"], "delta": "0:00:00.138730", "end": "2018-09-11 09:52:25.943765", "msg": "non-zero return code", "rc": 1, "start": "2018-09-11 09:52:25.805035",
"stderr": "error: cannot use --enable when the cluster is already initialized, use cluster add instead", 
"stderr_lines": ["error: cannot use --enable when the cluster is already initialized, 
use cluster add instead"], "stdout": "", "stdout_lines": []}
```

We need an additional test to add the feature if the cluster is already initialized as mentioned OR to delete systematically the cluster .... To be checked.

**Comment by geoand** [comment]
Yes, this does happen when you try to start the cluster again after it has been started (same as if you run `oc cluster up --enable ...` locally).

The easiest solution would be to ensure that everything is cleaned out before starting the cluster.
I will see however if we can check if the cluster has already previously been started

**Comment by cmoulliard** [comment]
I did some modifications concerning this ticket and added a new task to do `enablement` post `oc cluster up` like also to remove the dir containing `clusterup config files`

**Comment by geoand** [comment]
Me too :)

**Comment by geoand** [comment]
Feel free to commit your changes, otherwise I will open a PR with my changes

**Comment by cmoulliard** [comment]
I can't recreate a new cluster as I get this error now 

```
TASK [cluster : Config to start the all in one Openshift cluster using command 'oc cluster up'] ***********************************************************************************************************************************************
ok: [192.168.99.50] => {
    "msg": "Parameters :  --base-dir=/var/lib/origin/openshift.local.clusterup --public-hostname=192.168.99.50 --routing-suffix=192.168.99.50.nip.io --server-loglevel=1"
}

TASK [cluster : Enable options] ***************************************************************************************************************************************************************************************************************
changed: [192.168.99.50]

TASK [cluster : Launch cluster] ***************************************************************************************************************************************************************************************************************
fatal: [192.168.99.50]: FAILED! => {"changed": true, "cmd": ["oc", "cluster", "up"], "delta": "0:00:01.695310", "end": "2018-09-11 10:44:24.838106", "msg": "non-zero return code", "rc": 1, "start": "2018-09-11 10:44:23.142796", "stderr": "error: OpenShift is already running", "stderr_lines": ["error: OpenShift is already running"], "stdout": "Getting a Docker client ...\nChecking if image openshift/origin-control-plane:v3.10 is available ...\nChecking type of volume mount ...\nDetermining server IP ...\nChecking if OpenShift is already running ...", "stdout_lines": ["Getting a Docker client ...", "Checking if image openshift/origin-control-plane:v3.10 is available ...", "Checking type of volume mount ...", "Determining server IP ...", "Checking if OpenShift is already running ..."]}
        to retry, use: --limit @/Users/dabou/Code/snowdrop/openshift-infra/ansible/playbook/cluster.retry
```

**Comment by geoand** [comment]
What is the exact sequence of commands you are using that results in this issue?

I will finish testing my changes and open a PR then.

**Comment by cmoulliard** [comment]
Sequence :  

- Created a new VM -> centos7
- Generate inventory -> `ansible-playbook playbook/generate_inventory.yml -e ip_address=192.168.99.50 -e type=simple`
- Executing `up` top of master branch -> `ansible-playbook -i inventory/simple_host playbook/cluster.yml --tags "up"`

Question : why do we have 2 consecutive `oc cluster up` commands

```
- name: Enable options
  command: "oc cluster up {{ openshift_up_options }}"
  register: clusterupout
  ignore_errors: yes

- name: Launch cluster
  command: "oc cluster up"
  register: clusterupout
```

**Comment by geoand** [comment]
When I was first testing this about a month ago, the cluster would not start with the first command (with the enabled options)

**Comment by geoand** [comment]
That sequence of commands works for me, however I am having trouble deleting the state of the cluster

**Comment by cmoulliard** [comment]
I think that this command should be enough with 3.10

```
- name: Enable options
  command: "oc cluster up {{ openshift_up_options }}"
  register: clusterupout
#  ignore_errors: yes

#- name: Launch cluster
#  command: "oc cluster up"
#  register: clusterupout

- name: Cluster up
  include_tasks: cluster_up.yml
```

**Comment by geoand** [comment]
I will try it again, but last time it did not work :)

**Comment by cmoulliard** [comment]
It works for me. I will replay all the scenario including also the new task to enable the features such as service broker, ...

**Comment by geoand** [comment]
I tried it and it did not work.

Can you please commit your changes so I can test them and see what is different?

**Comment by geoand** [comment]
When I don't include `command: "oc cluster up"` then all that happens is that `/root/openshift.local.clusterup` get's created.

`docker ps` shows no containers running

**Comment by geoand** [comment]
I have committed my changes to: https://github.com/snowdrop/openshift-infra/tree/%23106

**Comment by cmoulliard** [comment]
We still have a problem to fix when we execute the  `clean` role

```
failed: [192.168.99.50] (item=/var/lib/origin/openshift.local.clusterup) => {"changed": false, "item": "/var/lib/origin/openshift.local.clusterup", "msg": "rmtree failed: [Errno 16] Device or resource busy: '/var/lib/origin/openshift.local.clusterup/openshift.local.volumes/pods/90fe3838-b59e-11e8-beef-080027e80dd4/volumes/kubernetes.io~secret/kube-dns-token-sb862'"}
```

**Comment by geoand** [comment]
My latest push to the `#106` branch should fix that

**Comment by cmoulliard** [comment]
I will push another change and then we could also try/merge your branch #106

**Comment by geoand** [comment]
Sounds good to me

**Comment by cmoulliard** [comment]
Retested completely with the tasks :  `up`, `down`, `clean` and `enable`. Can you have a look please ?

Something to be fixed which occurs when by example we enable cluster role `ansible-playbook -i inventory/simple_host playbook/post_installation.yml -e openshift_admin_pwd=admin --tags "enable_cluster_role"
` is 

```
TASK [openshift_env : assert] **************************************************************************************************************************************************************************************************************
fatal: [192.168.99.50]: FAILED! => {
    "assertion": "master_config_dir != ''", 
    "changed": false, 
    "evaluated_to": false, 
    "msg": "Unable to correctly determine the Openshift environment"
}
```

**Comment by geoand** [comment]
@cmoulliard That is not an error, it is needed to determine the environment that the playbook is running in

**Comment by geoand** [comment]
If it doesn't work, it means that the environment was not correctly set

**Comment by geoand** [comment]
I will check out the changes you have pushed and if anything needs to be fixed :0

**Comment by cmoulliard** [comment]
Something is failing as user is not promoted to `cluster-admin` -> https://gist.github.com/cmoulliard/e48c5e17461bf93ecc3e60a21008c85a

**Comment by geoand** [comment]
It most likely has to do with the change to the location where `oc cluster up` writes it's configuration. I am running your changes now so I will fix whatever issues I find

**Comment by cmoulliard** [comment]
I'm getting a very strange error now when I execute `enable` task

```
{
  "changed": true,
  "cmd": [
    "oc",
    "cluster",
    "add",
    "--base-dir=/var/lib/origin/openshift.local.clusterup",
    "service-catalog"
  ],
  "delta": "0:00:30.192690",
  "end": "2018-09-11 12:01:43.932546",
  "item": "service-catalog",
  "msg": "non-zero return code",
  "rc": 1,
  "start": "2018-09-11 12:01:13.739856",
  "stderr": "I0911 12:01:13.880075   11752 interface.go:26] Installing \"openshift-service-catalog\" ...\nI0911 12:01:13.900350   11752 apply_template.go:83] Installing \"service-catalog\"\nE0911 12:01:43.927582   11752 interface.go:34] Failed to install \"openshift-service-catalog\": failed to install \"service-catalog\": cannot create container using image openshift/origin-cli:v3.10; caused by: cannot create container using image openshift/origin-cli:v3.10\nI0911 12:01:43.927623   11752 interface.go:41] Finished installing \"openshift-service-catalog\"\nError: failed to install \"service-catalog\": cannot create container using image openshift/origin-cli:v3.10; caused by: cannot create container using image openshift/origin-cli:v3.10\n\n\nUsage:\n  oc cluster add [flags]\n\nExamples:\n  # Add service catalog\n  oc cluster add service-catalog\n  \n  # Add template service broker to a different basedir\n  oc cluster add --base-dir=other/path template-service-broker\n\nOptions:\n      --base-dir='': Directory on Docker host for cluster up configuration\n      --image='openshift/origin-${component}:${version}': Specify the images to use for OpenShift\n      --server-loglevel=0: Log level for OpenShift server\n\nUse \"oc options\" for a list of global command-line options (applies to all commands).",
  "stderr_lines": [
    "I0911 12:01:13.880075   11752 interface.go:26] Installing \"openshift-service-catalog\" ...",
    "I0911 12:01:13.900350   11752 apply_template.go:83] Installing \"service-catalog\"",
    "E0911 12:01:43.927582   11752 interface.go:34] Failed to install \"openshift-service-catalog\": failed to install \"service-catalog\": cannot create container using image openshift/origin-cli:v3.10; caused by: cannot create container using image openshift/origin-cli:v3.10",
    "I0911 12:01:43.927623   11752 interface.go:41] Finished installing \"openshift-service-catalog\"",
    "Error: failed to install \"service-catalog\": cannot create container using image openshift/origin-cli:v3.10; caused by: cannot create container using image openshift/origin-cli:v3.10",
    "",
    "",
    "Usage:",
    "  oc cluster add [flags]",
    "",
    "Examples:",
    "  # Add service catalog",
    "  oc cluster add service-catalog",
    "  ",
    "  # Add template service broker to a different basedir",
    "  oc cluster add --base-dir=other/path template-service-broker",
    "",
    "Options:",
    "      --base-dir='': Directory on Docker host for cluster up configuration",
    "      --image='openshift/origin-${component}:${version}': Specify the images to use for OpenShift",
    "      --server-loglevel=0: Log level for OpenShift server",
    "",
    "Use \"oc options\" for a list of global command-line options (applies to all commands)."
  ],
  "stdout": "Adding service-catalog ...",
  "stdout_lines": [
    "Adding service-catalog ..."
  ]
}
```

**Comment by geoand** [comment]
I pushed a change to master that should fix your issue with `cluster-admin` and other tasks

**Comment by geoand** [comment]
There is still a problem with the location of the config, I need to look into it
